### PR TITLE
Clean output

### PR DESCRIPTION
**Pull request by flying-sheep**
Addresses https://github.com/bskinn/sphobjinv/issues/131#issuecomment-716789838

This results in messages being shown without clobbering stdout and reduces duplicated code.

See here: the first lines are stderr, the formatted JSON is the stdout after going though `jq`

```console
$ python -m sphobjinv convert json -u https://rpy2.github.io/doc/latest/html/ - | jq
 
No inventory at provided URL.
Attempting "https://rpy2.github.io/doc/latest/html/objects.inv" ...
Remote inventory found.
 
{
  "project": "rpy2",
  "version": "3.3.1",
  "count": 294,
  "0": {
  ...
}
```

### CODE REVIEW

File: src/sphobjinv/cmdline.py [hunk 1]
```diff
@@ -774,53 +741,55 @@ def do_convert(inv, in_path, params):
...
-            write_json(inv, out_path, expand=params[EXPAND], contract=params[CONTRACT])
+            write_json(inv, file, expand=params[EXPAND], contract=params[CONTRACT])
+    except BrokenPipeError:
+        pass
     except Exception as e:
         selective_print("\nError during write of output file:", params)
         selective_print(err_format(e), params)
         sys.exit(1)
+    finally:
+        file.close()
```
Comment (bskinn) [inline src/sphobjinv/cmdline.py:782]: I guess calling `.close()` on `sys.stdout.buffer` doesn't cause a problem for the next executable, trying to read the piped data?

**Comment by bskinn** [comment]
I do like how this revision to the stream writing tightens up the code. However, it totally obliterated the test suite, so I'd rather not merge it in the short term. Odds are my hacky stdio management in the test suite also really needs to be cleaned up, and would 

I'll fix `selective_print()` and cut another alpha soon.

File: src/sphobjinv/cmdline.py (hunk 1 above)
Comment (flying-sheep) [reply src/sphobjinv/cmdline.py:782]: No, try it! You can add any pipe here:

```console
$ python -c 'import; b = sys.stdout.buffer; b.write(b"1"); b.close()'
```

**Comment by flying-sheep** [comment]
Yeah, the `selective_print` thing is just a one line fix anyway!

One could also thing about e.g. adding the final url to the JSON output instead of (just) logging it. That way it would be available programmatically.

**Comment by bskinn** [comment]
Oooh, I like that. I even already have a spot in the JSON schema suitable for metadata like that. Will add!

**Comment by bskinn** [comment]
v2.1a2 is up on PyPI -- the stdout content should be clean, and for JSON output (stdout and written file) that was drawn from a URL, the URL where the inventory was found is stored in `result.metadata.url`.

Let me know how it does for you!

**Comment by bskinn** [comment]
Nodnod, I'm really glad I found it. Note also the `.. program-output::` directive, which doesn't echo the command.

There are a number of useful options; see [docs](https://sphinxcontrib-programoutput.readthedocs.io/en/latest/).

**Pull request closed without merging**